### DISCUSSION TITLE
Generating classes during compalation

### DISCUSSION THREAD

**Discussion by sktrendla** [symfony/symfony#53568] [Q&A] [answered]
Hello,

I was trying to implement a simple feature flag system with the help of the dependency injection component.  I was hoping I could find documentation on how to generate classes during compilation but I failed with that. The compiler passes seem to have the ability to generate classes during compilation and use them in dependency injection, the example would be the lazy services.

The questions that I have is the following:

- Is there a built in way of handling class generation during compilation in Symfony as a framework that can be built on?
- If there is a way to do that is there documentation available for it?
- Is the approach to generate classes this way a generally bad with Symfony?

A question was asked about a similar topic #53274  earlier but this one is strictly about the generation mechanism for classes instead of their handling afterwards.

**Comment by MatTheCat** [comment]
Hello,

I don’t understand how a feature flag system would need to generate classes during compilation; this looks like a [X/Y problem](https://en.wikipedia.org/wiki/XY_problem) to me, and you probably should consider another way.

**Reply by sktrendla** [reply]
Hi,

thanks for the response. I admit the example might be a bad one and the example probably should be implemented  in a different way. A more abstract or technical version of what I was attempting to do in my example would be the following.

I was attempting to generate boilerplate code that had the following properties:

1. It's type must be compatible with the class it would decorate.
2. On every function call on it must make a call to a third service without using any information from the called function's parameters, and then would forward the call to the decorated service and return with it's response. 
3. It must be always be up to date with the decorated service's type so no functions would be missed.

To me these properties suggested that the code should not be created or maintained by humans, nor should it be directly visible to them.
The natural conclusion seems to be some simple classes generated during compilation and using them with the help of the dependency injection system.

The original example in this case is almost irrelevant, it was just an attempt at demonstrating why I'd need to generate classes.

**Comment by ghost** [comment, marked as answer]
I think that the class generation (i.e. a physical file) is something you have to do on your own. I don't think Dependency Injection component has an API for such things. It does some class generation but it's just for the component's own use.
